### PR TITLE
NIFI-1614 File Identity Provider implementation

### PR DESCRIPTION
**Pull request by jvwing**
This implementation of IdentityProvider uses a local XML file to store usernames with bcrypt-hashed passwords.  The XML file mechanism is similar to that used by the FileAuthorizationProvider.  Requires that NiFi has been configured for secure communication with SSL.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/login-identity-providers.xml [hunk 1]
```diff
@@ -89,4 +89,28 @@
...
+        through authorized-users.xml.
+    -->
+    <!-- To enable the file-identity-provider remove 2 lines. This is 1 of 2.
+    <provider>
+        <identifier>file-identity-provider</identifier>
+        <class>org.apache.nifi.authentication.file.FileIdentityProvider</class>
+        <property name="Credentials File">credentials.xml</property>
+        <property name="Authentication Expiration">12 hours</property>
+    </provider>
+    To enable the file-identity-provider remove 2 lines. This is 1 of 2. -->
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/login-identity-providers.xml]: Is this actually line 2 of 2? If not, where is the other line to be removed?

File: nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    @Override
+    public final void onConfigured(final LoginIdentityProviderConfigurationContext configurationContext) throws ProviderCreationException {
+        final Map<String, String> configProperties = configurationContext.getProperties();
+        for (String propertyKey : configProperties.keySet()) {
+            String propValue = configProperties.get(propertyKey);
+            logger.debug("found property '{}': '{}'", propertyKey, propValue);
+        }
+
+        credentialsFilePath = configProperties.get(PROPERTY_CREDENTIALS_FILE);
+        if (credentialsFilePath == null) {
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: Can this return an empty String? `StringUtils.isEmpty()` may be a better check here than just `== null`.

File: nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    @Override
+    public final AuthenticationResponse authenticate(final LoginCredentials credentials) throws InvalidLoginCredentialsException, IdentityAccessException {
+        final String loginUsername = credentials.getUsername();
+        final String loginPassword = credentials.getPassword();
+        AuthenticationResponse authResponse = null;
+
+        try {
+            UserCredentialsList credentialsList = loadCredentialsList(credentialsFilePath);
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: This will try to read from the file system and parse the XML schema every time `authenticate` is called. Does it make more sense to cache the credentials in memory (thus requiring a restart if new users are added/users are removed, etc.)?

File: nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            List<UserCredentials> userCredentials = credentialsList.getUser();
+            for (UserCredentials userCreds : userCredentials) {
+                final String storedUsername = userCreds.getName();
+                if (storedUsername.equals(loginUsername)) {
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: The username is case-sensitive in this case. Is this necessary, or can `equalsIgnoreCase` be used?

Comment (alopresto) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: I don't see anywhere that a user can register, so the credential store on the filesystem can only currently be modified manually by someone with `write` access to that file. However, it doesn't appear that there are any safeguards on the data contained therein -- no unique constraint on the usernames, so in the following scenario, what would happen?

```xml
<credentials>
    <user name="user1" passwordHash="bcrypt(password)" />
    <user name="user1" passwordHash="bcrypt(real&Strong&Password)" />
</credentials>
```

If a login request for `user1`, `password` came in, would it succeed? What about `user1`, `real&Strong&Password`?

What is the mechanism for an administrator adding a user to the credential store? They must write a bespoke Java method to hash the user-provided password and put the resulting hash into the XML file by hand? This is the same for password resets and updates?

There is no default  `credentials.xml` file provided with this NAR but should an empty one be provided with default file system permissions locked down to further indicate that anyone with `read` access to this file can perform brute-force or pattern-matching hash analysis against the credentials store, and anyone with `write` access can simply replace a user's password hash and then gain access to the system with their account?

I understand this is not designed for a production system but it still seems dangerous to release as is, because well-intentioned but short-on-time people will deploy this to production.

Comment (alopresto) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: The XSD does not require that the username is non-null, so this could throw a NPE.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/login-identity-providers.xml (hunk 1 above)
Comment (jvwing) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/login-identity-providers.xml]: Yes, thanks, it should say 2 of 2.  I will change that.

File: nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java (hunk 2 above)
Comment (jvwing) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: Good point, I will add a unit test for the empty string case and use StringUtils.isEmpty() for the check.  

Validating the existence and access to the file depends on the choice made about startup or runtime file access pattern, I'll comment more on that below. It certainly seems reasonable to at least log a WARN message about the missing file here.  I will at least do that.

File: nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java (hunk 3 above)
Comment (jvwing) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: I did consider this, and I decided on the repetitive read, no-restart-required design.  My assumption depends on there being relatively few users and a relatively generous token expiration period.  With a 12 hour - 24 hour token expiration period, it did not seem to matter if we read the file 100 times a day to log in 100 users.  And how many NiFi installations have 100 users?  And of those 100+ user installations, how many are not already down the LDAP or X.509 path?  I decided administrative flexibility without restarting would be a nice and inexpensive benefit, fitting with the theme of an 'easier' identity scheme.

Do you have a strong opinion?  What working assumptions about numbers of users, expiration periods, and login frequency would you follow?

File: nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java (hunk 4 above)
Comment (jvwing) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: Case-sensitivity on Usernames - Thank you, I'm afraid I had not thought that through properly.  I believe this should be consistent with the FileAuthorizationProvider (at least), which appears to be case-insensitive.  I'll research more and update appropriately.

Comment (jvwing) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: I will add an XSD unique constraint on the usernames, and a unit test for the validation.  It looks like XSD constraints do not (easily?) do case-insensitive unique comparisons, so it won't be foolproof, but it may help prevent some accidents.

Comment (jvwing) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: > ```
> <credentials>
>     <user name="user1" passwordHash="bcrypt(password)" />
>     <user name="user1" passwordHash="bcrypt(real&Strong&Password)" />
> </credentials>
> ```

> If a login request for user1, password came in, would it succeed? What about user1, real&Strong&Password?

I believe the unique validation constraint discussed above would prevent this sample from running.  Additionally, `authenticate()` will only match one username, and try one password hash, defined as whichever it encountered first in the list.  I do not sort the list, just accept the default ordering from JAXB unmarshalling.  Is that bad?

**Comment by jvwing** [comment]
@alopresto, thanks for kick-starting the discussion with your feedback, these are great topics.  I address some of the concerns you mention below, and ask some questions back to you at the end.

#### Incremental Delivery
I propose this PR as a viable first step on a potentially long path.  I believe this increment can be useful to pilot the concept on its own, leaving further enhancements pending feedback and expressed interest.


#### Credentials File
I decided not to provide a sample credentials XML file.  If included, it would most naturally fit with other configuration files in the `conf` directory.  But the default permissions on the conf directory files are not appropriate for this credentials file.  I felt that including the file at that location would lead to it simply being left there.  On the other hand, creating an entirely new folder or permission might be presumptuous for a controversial and unproven feature.

After carefully considering these factors, I chose the lazy and cowardly way out by just documenting what the file should look like, and leaving it up to the user to find a home for it.


#### User Management Tool
I provide no tool in this PR for generating password hashes. I simply document it as "bcrypt 2a, 10 rounds".  bcrypt is a standard of sorts, and there are many libraries and some command-line utilities available for it.

I am absolutely open to providing a tool, bundled in NiFi or sold separately.  I did not find similar command-line utilities shipped with NiFi, and again, it seemed presumptuous to build out a structure for that at this time.  Am I missing them?

There are also other possible hashing algorithm and tooling combinations.  We might consider SHA256 or another algorithm that might have better default installation support.  I wasn't real happy with the options OpenSSL provided, MD5 is widely derided for passwords, but bcrypt fits the bill and was conveniently included in the Spring package.  I'm open to others, and an obvious future expansion would be to permit several.


#### Comparability with Other Providers
I considered that NiFi provides no tool for generating certificates, does not include default keystore or truststore files, nor suggest their locations and what permissions you should or should not have on them.  This is an imperfect comparison, since this intends to be simpler, but perhaps a reasonable choice for an initial release.


#### General Wisdom Of Doing This At All
This identity provider is opt-in.  As you point out, it currently requires determined configuration and admin work to get it running given that I provide neither tools nor a credentials file.


#### Questions
* What is required to make this viable?
* Is there a better medium than bcrypt that combines widespread tool support with decent encryption.
* Are we open to including a command-line user admin tool?
* Are we open to including a sample credentials file?  Where would you recommend it go?
* Are we open to documenting this identity provider on the front-page of the Admin Guide alongside X.509 and LDAP?  Where else should I do so?

**Comment by alopresto** [comment]
@jvwing James, you've obviously put effort and thought into this and I didn't mean to deride it. I am working to get the Kerberos authentication completed for the release, but I will re-visit this as soon as I finish up with that. 

I think historically, user management and authentication has been "certificates or bust" from the NiFi point of view, delegating that responsibility to tools more focused on that concern. LDAP integration was just recently added, and Kerberos is obviously yet to be delivered. 

Off the top of my head, Bcrypt is definitely a strong candidate for the password hashing (single-iteration SHA-256 is essentially the same as plaintext), the certificate support is fairly well-documented in the [Admin Guide](https://nifi.apache.org/docs/nifi-docs/html/administration-guide.html#security-configuration), keystores are not user-editable flat files and are password-protected by default, and there are substantial third-party resources surrounding the documentation of "securing a NiFi instance" from various organizations.

File: nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java (hunk 3 above)
Comment (alopresto) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: I understand the trade-off between convenience and performance. While I agree that installations with hundreds of users will likely be using a more robust user authentication mechanism, we should be aware there will be a performance degradation here. Pre-emptive optimization is unnecessary for now.

Comment (apiri) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: An alternative might be to leverage the database that is currently used for users. Potentially read at startup and load those entries to at least avoid the full hit on each check.  Would require some adaptation and extension, but a good bit of the effort is already present.

Comment (jvwing) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/FileIdentityProvider.java]: How about loading and saving the list in memory, but checking the last modified date on the file before each login to ensure that it is current?  This would allow administration on the fly, while skipping the slowest XML parsing step.

**Comment by joewitt** [comment]
@jvwing @alopresto Have you had a chance to re-engage on this?  It seems like a reasonable easy-path option for folks just wanting to use some simple/local username and password based setup.  My responses to the questions James posed:

What is required to make this viable?
- This discussion appears on track

Is there a better medium than bcrypt that combines widespread tool support with decent encryption.
- Sounds like you and Andy both see it as a good option.

Are we open to including a command-line user admin tool?
- In my opinion we should be consistent that administrative actions occur by editing files on the command line in the less optimal case and interacting through a designed/intentional UX in the best case.  We should strive to move away from config file based options and move fully towards service/REST API driven approaches.  These will serve us better in clustered/cloud type environments as well.

Are we open to including a sample credentials file? Where would you recommend it go?
- Absolutely.  In conf directory like the others of its type.  I think an argument could be made to have this username/password driven mode be the default.

Are we open to documenting this identity provider on the front-page of the Admin Guide alongside X.509 and LDAP? Where else should I do so?
- We must do so.  We should fully embrace this as an option and document what it is good for and not good for.  Our current default of having no authentication at all is what we should be working to eliminate.  I think this offers us a good first step to do that.

**Comment by jvwing** [comment]
I have been working (slowly) on the suggested improvements for performance and a reference CLI for basic admin operations.  I haven't figured out an elegant way of packaging a CLI-executable class in a NAR file, but I agree there should be some most basic tool available to generate proper hashes, or at least demonstrate how to do so.

I will add a sample config file and documentation.  

I do not propose to add any UI features within this ticket/PR.  However, I'm curious how you think that might work.  Can plugin NARs can add to the web API and UI, or would that imply tighter coupling with the core features?  For the moment, I see this as a more peripheral and optional plugin.

**Comment by jvwing** [comment]
I rebased the commits on the master branch to resolve conflicts and use the updated LoginIdentityProvider interface and Administrator's Guide content.  I apologize if it complicates reviewing.  Changes include:

- Improved performance by only reloading the credentials data if the file has been modified
- Provided a command-line utility reference implementation
- Added documentation to the Administrator's Guide
- Included a sample login-credentials.xml file to the conf directory

File: nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/CredentialsCLI.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+
+/**
+ * <p>Command-line interface for working with a {@link CredentialsStore}
+ * persisted as an XML file.</p>
+ *
+ * <p>Usage:</p>
+ * <ul style="list-style-type:none">
+ *   <li>list credentials.xml</li>
+ *   <li>add credentials.xml admin password</li>
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/CredentialsCLI.java]: Accepting the raw password on the command line will mean that it is persisted in the terminal history and available to any other processes running. It is more secure to use [Console#readPassword()](https://docs.oracle.com/javase/7/docs/api/java/io/Console.html#readPassword%28%29)

Comment (jvwing) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/CredentialsCLI.java]: Thanks, I'll try that method.

File: nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/CredentialsStore.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import javax.xml.bind.Unmarshaller;
+import javax.xml.bind.ValidationEvent;
+import javax.xml.bind.ValidationEventHandler;
+import javax.xml.transform.stream.StreamSource;
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+
+import org.apache.nifi.authentication.file.generated.ObjectFactory;
```
Comment (alopresto) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/CredentialsStore.java]: I'm getting a number of issues building the project with these generated classes. Why do they need to be generated?

**Comment by alopresto** [comment]
I think my earlier comments may have been unclear or ambiguous. I do not believe we need a full command-line interface for modifying the configuration file, as hand-editing the files is the existing norm. I simply meant that the process of protecting a raw password with bcrypt is not an "in-head" operation for most instance admins, so we should provide a utility to perform that operation. 

While I very much respect the effort that went into the supporting infrastructure, I think it is overkill and not consistent with the global approach.

File: nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/CredentialsStore.java (hunk 6 above)
Comment (jvwing) [inline nifi-nar-bundles/nifi-iaa-providers-bundle/nifi-file-identity-provider/src/main/java/org/apache/nifi/authentication/file/CredentialsStore.java]: They are used to serialize and deserialize the XML credentials file.  What kind of issues are you experiencing?

**Comment by jvwing** [comment]
What would you recommend for this pull request?  No utility?  A simpler hashing utility?

**Comment by alopresto** [comment]
My recommendation would be a command line utility (could be a shell script wrapping a JAR) which is  invoked without arguments, provides a secure prompt for entering a raw password, and returns the hashed password in _bcrypt_ format. 

I think the Java code necessary to read from the XML configuration file, accept a username and password entered in the UI form field, verify the credentials are correct, and return a JWT authentication token should remain. 

The administrative functionality of the utility (reading and writing the XML file) does not seem necessary to me. 

Again, I understand the desire for this from a proof-of-concept and quick-deploy mindset, but I'll reiterate my feeling that this lightweight option will become an accidental de facto default because it requires less effort to deploy than LDAP, Kerberos, or client certificate authentication. Currently, I believe those alternatives are all more robust and secure than this approach:

- Each is a widely supported approach with many eyes on them, making discovery of vulnerabilities more likely
- Each protects the knowledge of the end users' passwords from the administrator, unlike this approach
- Each scales without the manual intervention of the administrator (authorization still requires manual intervention)

I appreciate the effort to move away from anonymous access as the default for the application. However, I believe there should be a balance here, as the default is very clear that no authentication is performed. Administrators must make an intentional effort to secure the instance with LDAP, Kerberos, or client certificates, all of which I acknowledge have a higher cost of entry than simple username/password authentication from a file store. My fear is that as this method is the lowest cost, it will be explained (incompletely) on a random blog article, become the quickest route for admins to get it running without fully reading the documentation or knowing its limitations, and then be deployed to a production environment. 

I understand these concerns are mostly intuitive and not examples I can point to with empirical evidence.

**Comment by jvwing** [comment]
I added a new CLI utility that just hashes passwords using Bcrypt.  For the moment, I left the rest of the code alone.  I'm still struggling with how to best package and distribute any command-line utility.  There are currently a couple of undesirable steps to running this: 

1. Build a standalone jar of nifi-file-identity-provider using `mvn compile assembly:single`
1. Run the standalone jar like `java -jar nifi-file-identity-provider-1.0.0-SNAPSHOT-jar-with-dependencies.jar`

I do not know how to run the class from within the NAR file.  It seems like we would need to get the module to compile to JAR, get the JAR in the lib folder, and then make a shell script to wrap the nice java syntax.  Any ideas on that?

**Comment by alopresto** [comment]
I'm away from my keyboard right now but I have always used the `maven:shade` plugin.

Andy LoPresto
alopresto@hortonworks.com<mailto:alopresto@hortonworks.com>
Sr. Member of Technical Staff
410.292.6331<tel:410.292.6331>
PGP Fingerprint: 9B51 8C0D A489 0D3C 2F5B  0F2B 125A 4E68 51BF 2B79

On May 3, 2016, at 12:48, James Wing <notifications@github.com<mailto:notifications@github.com>> wrote:


I added a new CLI utility that just hashes passwords using Bcrypt. For the moment, I left the rest of the code alone. I'm still struggling with how to best package and distribute any command-line utility. There are currently a couple of undesirable steps to running this:

  1.  Build a standalone jar of nifi-file-identity-provider using mvn compile assembly:single
  2.  Run the standalone jar like java -jar nifi-file-identity-provider-1.0.0-SNAPSHOT-jar-with-dependencies.jar

I do not know how to run the class from within the NAR file. It seems like we would need to get the module to compile to JAR, get the JAR in the lib folder, and then make a shell script to wrap the nice java syntax. Any ideas on that?

-

**Comment by trixpan** [comment]
@jvwing - would you be able to rebase this PR?

**Comment by jvwing** [comment]
Sorry, I have not kept this up to date with changes in master.  I will close this PR for now.  It may be a better fit for a future extension repository.
